### PR TITLE
MINOR: Include the inner exception stack trace when re-throwing an exception

### PR DESCRIPTION
**Pull request by divijvaidya**
# Problem: 
While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

# Change
When re-throwing an exception, we make sure to include the stack trace. Otherwise pertinent debug information is lost.

### CODE REVIEW

**Comment by divijvaidya** [comment]
@dengziming @hachikuji kindly review.

Reviewer(s) please note that all tests failing for this CI are known to be flaky and are not related to this code change.

Test 1 - testListenerConnectionRateLimitWhenActualRateAboveLimit - fix is pending PR review https://github.com/apache/kafka/pull/12045

Test 2 - testTopicIdUpgradeAfterReassigningPartitions - fix is pending PR review https://github.com/apache/kafka/pull/11687

Test 3 - testSnapshotOnlyAfterConfiguredMinBytes - (possible) fix is pending PR review https://github.com/apache/kafka/pull/12224

Test 4 - Created a ticket at https://issues.apache.org/jira/browse/KAFKA-13951

**Review by Kvicii — APPROVED** [review]
LGTM.

**Comment by divijvaidya** [comment]
@hachikuji Please review. This should be a small one.

**Review by dengziming — APPROVED** [review]
Can this kind of problem be caught by spotbugs? manual checking is error prone.

**Comment by divijvaidya** [comment]
> Can this kind of problem be caught by spotbugs? manual checking is error prone.

Agreed @dengziming but unfortunately spotbugs isn't catching such errors.

**Comment by divijvaidya** [comment]
> Can this kind of problem be caught by spotbugs? manual checking is error prone.



> Can this kind of problem be caught by spotbugs? manual checking is error prone.



> Can this kind of problem be caught by spotbugs? manual checking is error prone.

Agreed @dengziming but unfortunately spotbugs isn't catching such errors.

**Comment by divijvaidya** [comment]
@mimaison perhaps you may want to look into this? This already has 2 approvals from non-committers.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java]: Have we checked that we're not introducing a security vulnerability in these security classes by including the original exception?

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: Does the underflow exception actually add any information? If I recall correctly, it does not.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/CompactArrayOf.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/CompactArrayOf.java]: I think it would be more helpful to include the class of `item` than chaining the exception here.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java]: `UnsupportedEncodingException` is thrown for 2 cases and in either of the two cases, the string being encoded is not printed in the exception message and hence, we don't leak any message information using the exception.

Case 1: for `UnsupportedCharsetException` in which case, only charset is printed in the string as per the implementation as below:
```
public class UnsupportedCharsetException
    extends IllegalArgumentException
{

    private static final long serialVersionUID = 1490765524727386367L;

    private String charsetName;

    /**
     * Constructs an instance of this class.
     *
     * @param  charsetName
     *         The name of the unsupported charset
     */
    public UnsupportedCharsetException(String charsetName) {
        super(String.valueOf(charsetName));
	this.charsetName = charsetName;
    }

    /**
     * Retrieves the name of the unsupported charset.
     *
     * @return  The name of the unsupported charset
     */
    public String getCharsetName() {
        return charsetName;
    }

}
```

Case 2: for `IllegalCharsetNameException` which again, just prints the name of the charset. See implementation as below:
```
public class IllegalCharsetNameException
    extends IllegalArgumentException
{

    private static final long serialVersionUID = 1457525358470002989L;

    private String charsetName;

    /**
     * Constructs an instance of this class.
     *
     * @param  charsetName
     *         The illegal charset name
     */
    public IllegalCharsetNameException(String charsetName) {
        super(String.valueOf(charsetName));
	this.charsetName = charsetName;
    }

    /**
     * Retrieves the illegal charset name.
     *
     * @return  The illegal charset name
     */
    public String getCharsetName() {
        return charsetName;
    }

}
```

Is there any other security risk that you are alluding to here? Adding the stack trace is beneficial here to quickly determine whether the failure is due to case 1 or case 2.

File: clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java]: You are right, it doesn't. I will revert this.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/CompactArrayOf.java
Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/CompactArrayOf.java]: I have fixed it as per your suggestion.

**Comment by divijvaidya** [comment]
@ijuma I have addressed your comments. This is ready for your review (pending test run).

**Review by mdedetrich — APPROVED** [review]
These changes seem good for me

**Comment by divijvaidya** [comment]
@tombentley  requesting code review bandwidth from you here. We already have 3 non-committer approvals on this one.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/secured/HttpAccessTokenRetriever.java]: Thanks for the explanation @divijvaidya , yes, I checked and confirmed it only output the charset name. So it's safe. @ijuma , any concern/comments here?

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the improvement.

**Comment by divijvaidya** [comment]
Rebased with the latest `trunk` to resolve merge conflicts.

**Comment by showuon** [comment]
I'm going to merge this PR this week if there's no more comments. Thanks.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java]: Is this similar to the CompactArrayOf? Generally, if we can provide the right information without having a long chain of exceptions, it's better.

Comment (divijvaidya) [reply clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java]: That's a good tip Ismael. Thank you! I have fixed this in the latest commit.

**Review by ijuma — APPROVED** [review]
LGTM, thanks.

**Pull request merged**
